rasa/utils/train_utils.py

```diff
@@ -342,6 +368,94 @@ def override_defaults(
+    if (
+        component_config[LOSS_TYPE] == MARGIN
+        and component_config[MODEL_CONFIDENCE] == SOFTMAX
+    ):
+        rasa.shared.utils.io.raise_warning(
+            f"Overriding defaults by setting {MODEL_CONFIDENCE} to "
+            f"{COSINE} as {LOSS_TYPE} is set to {MARGIN} in the configuration."
+        )
+        component_config[MODEL_CONFIDENCE] = COSINE
+    return component_config
+
+
+def validate_configuration_settings(component_config: Dict[Text, Any]) -> None:
+    """Performs checks to validate that combination of parameters in the configuration are correctly set.
+
+    Args:
+        component_config: Configuration to validate.
+    """
+    _check_loss_setting(component_config)
+    _check_confidence_setting(component_config)
+    _check_similarity_loss_setting(component_config)
+
+
+def _check_confidence_setting(component_config: Dict[Text, Any]) -> None:
+    if component_config[MODEL_CONFIDENCE] == SOFTMAX:
+        rasa.shared.utils.io.raise_warning(
+            f"{MODEL_CONFIDENCE} is set to `softmax`. It is recommended "
+            f"to set it to `cosine`. It will be set to `cosine` by default, "
+            f"Rasa Open Source 3.0 onwards.",
+            category=FutureWarning,
+        )
+        if component_config[LOSS_TYPE] not in [SOFTMAX, CROSS_ENTROPY]:
+            raise InvalidConfigException(
+                f"{LOSS_TYPE}={component_config[LOSS_TYPE]} and "
+                f"{MODEL_CONFIDENCE}={SOFTMAX} is not a valid "
+                f"combination. You can use {MODEL_CONFIDENCE}={SOFTMAX} "
+                f"only with {LOSS_TYPE}={CROSS_ENTROPY}."
+            )
+        if component_config[SIMILARITY_TYPE] not in [INNER, AUTO]:
+            raise InvalidConfigException(
+                f"{SIMILARITY_TYPE}={component_config[SIMILARITY_TYPE]} and "
+                f"{MODEL_CONFIDENCE}={SOFTMAX} is not a valid "
+                f"combination. You can use {MODEL_CONFIDENCE}={SOFTMAX} "
+                f"only with {SIMILARITY_TYPE}={INNER}."
+            )
+
+
+def _check_loss_setting(component_config: Dict[Text, Any]) -> None:
+    if not component_config[CONSTRAIN_SIMILARITIES] and component_config[LOSS_TYPE] in [
+        SOFTMAX,
+        CROSS_ENTROPY,
+    ]:
+        rasa.shared.utils.io.raise_warning(
+            f"{CONSTRAIN_SIMILARITIES} is set to `False`. It is recommended "
+            f"to set it to `True` when using cross-entropy loss. It will be set to `True` by default, "
+            f"Rasa Open Source 3.0 onwards.",
+            category=FutureWarning,
+        )
+
+
+def _check_similarity_loss_setting(component_config: Dict[Text, Any]) -> None:
+    if (
+        component_config[SIMILARITY_TYPE] == COSINE
+        and component_config[LOSS_TYPE] == CROSS_ENTROPY
+        or component_config[SIMILARITY_TYPE] == INNER
+        and component_config[LOSS_TYPE] == MARGIN
+    ):
+        rasa.shared.utils.io.raise_warning(
```

Comment:
```suggestion
        rasa.shared.utils.io.raise_deprecation_warning(
```

Reply:
Is this a regular warning or a deprecation warning? It's not deprecated it should be `UserWarning`

Reply:
It's a regular warning, changing it to `UserWarning`

Reply:
Do you have a good pointer on when to use `UserWarning` v/s `FutureWarning`. I seem to be missing it always. My understanding was that if we are describing a behaviour that is expected to take place in the future, it should be `FutureWarning`.

Reply:
Anything which is just incorrect / not suggested behavior: `UserWarning`
If something is deprecated and will be removed in the future: `FutureWarning`

Does that make sense?

> a behaviour that is expected to take place in the future,

What do you mean by that?

Reply:
Mhm, I think I get your troubles now based on [this](https://github.com/RasaHQ/rasa/pull/7616#discussion_r572203976) comment.

You mean that the parameter isn't deprecated and won't be deprecated, but the default will change, right?

Reply:
then I'd say it's a regular `UserWarning`

Reply:
> mean that the parameter isn't deprecated and won't be deprecated, but the default will change, right?
Yes, alright then all of these are `UserWarning`. Thanks 👍